homeassistant/components/sonos/speaker.py

```diff
@@ -0,0 +1,234 @@
+    DOMAIN as SONOS_DOMAIN,
+    PLATFORMS,
+    SCAN_INTERVAL,
+    SEEN_EXPIRE_TIME,
+    SONOS_CONTENT_UPDATE,
+    SONOS_DISCOVERY_UPDATE,
+    SONOS_ENTITY_CREATED,
+    SONOS_ENTITY_UPDATE,
+    SONOS_GROUP_UPDATE,
+    SONOS_MEDIA_UPDATE,
+    SONOS_PLAYER_RECONNECTED,
+    SONOS_PROPERTIES_UPDATE,
+    SONOS_SEEN,
+    SONOS_STATE_UPDATED,
+    SONOS_VOLUME_UPDATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SonosSpeaker:
+    """Representation of a Sonos speaker."""
+
+    def __init__(self, hass: HomeAssistant, soco: SoCo):
+        """Initialize a SonosSpeaker."""
+        speaker_info = soco.get_speaker_info(True)
+
+        self._is_ready: bool = False
+        self._subscriptions: list[SubscriptionBase] = []
+        self._poll_timer: Callable | None = None
+        self._seen_timer: Callable | None = None
+        self._seen_dispatcher: Callable | None = None
+        self._entity_creation_dispatcher: Callable | None = None
+        self._platforms_ready: set[str] = set()
+
+        self.hass: HomeAssistant = hass
+        self.soco: SoCo = soco
+
+        self.mac_address = speaker_info["mac_address"]
+        self.model_name = speaker_info["model_name"]
+        self.version = speaker_info["software_version"]
+        self.zone_name = speaker_info["zone_name"]
+
+    def setup(self) -> None:
+        """Run initial setup of the speaker."""
+        self._entity_creation_dispatcher = dispatcher_connect(
+            self.hass,
+            f"{SONOS_ENTITY_CREATED}-{self.soco.uid}",
+            self.async_handle_new_entity,
+        )
+        self._seen_dispatcher = dispatcher_connect(
+            self.hass, f"{SONOS_SEEN}-{self.soco.uid}", self.async_seen
+        )
+        dispatcher_send(self.hass, SONOS_DISCOVERY_UPDATE, self)
+
+    async def async_handle_new_entity(self, entity_type: str) -> None:
+        """Listen to new entities to trigger first subscription."""
+        self._platforms_ready.add(entity_type)
+        if self._platforms_ready == PLATFORMS:
+            await self.async_subscribe()
+            self._is_ready = True
+
+    @callback
+    def async_write_entity_states(self) -> bool:
+        """Write states for associated SonosEntity instances."""
+        async_dispatcher_send(self.hass, f"{SONOS_STATE_UPDATED}-{self.soco.uid}")
+
+    @property
+    def available(self) -> bool:
+        """Return whether this speaker is available."""
+        return self._seen_timer is not None
+
+    @property
+    def device_info(self) -> dict[str, Any]:
+        """Return information about the device."""
+        return {
+            "identifiers": {(SONOS_DOMAIN, self.soco.uid)},
+            "name": self.zone_name,
+            "model": self.model_name.replace("Sonos ", ""),
+            "sw_version": self.version,
+            "connections": {(dr.CONNECTION_NETWORK_MAC, self.mac_address)},
+            "manufacturer": "Sonos",
+            "suggested_area": self.zone_name,
+        }
```

Comment:
It would be cleaner to move this into the base entity since nothing in this class cares about it